Multiple Input sources

shashwathi opened the pull request:
Fixes #392 

## Proposed Changes

  * Introduce new key called `sources` 
  * Introduce `name` key for `sources` 
  * Continue the behavior if build has `source` provided. 
  * Functionality of `name` key: `source` has `name` key and this field is used to determine the mounting folder name.
For example if 2 git sources are defined with name `foo` and `bar`. This will result in following folders under `workspace`

```
sources:
- name: foo
  git: ...
- name: bar
   git: ....
``` 
`/workspace/foo`
`/workspace/bar`

Why: This will allow build to define same input source under different namespace. Simple use case would be build can have 2 git sources as inputs and steps could include check for merge without conflicts between `develop` and `master`.

Internal changes
* Step names now include the source name
```
build-step-git-source-foo
build-step-git-source-bar
```
Gotchas
 * Build cannot have both source and sources. 
 * `source` included support for `subPath` but `sources` does not support subpath yet for multiple sources. If more than 2 sources have `subPath` defined, its invalid build in this PR. 
Why: All inputs are mounted on same volume and cannot set multiple subpath for a volume mount.\
Future enhancement: Build can be more intelligent to fetch repo and check subPath exists in future and mount only that. (Not in this PR) 

**Release Note**

```
`sources` will take array of sources for build.
```

<!--
Request Prow to automatically lint any go code in this PR:

/lint
-->
cc @ImJasonH @mattmoor

mattmoor commented:
I think @fejta may be interested in this.

shashwathi commented:
/assign @ImJasonH

vdemeester commented on cmd/git-init/main.go:
```diff
@@ -66,10 +68,30 @@ func main() {
+	if *name != "" {
+		// create dir name
+		path := filepath.Join(dir, *name)
+		if _, err := os.Stat(path); os.IsNotExist(err) {
```
any reason for not using `os.MkdirAll` that would do the same almost ?

shashwathi replied:
I did not want to use `os.MkdirAll` because I expect the `workspace` dir to already exist and `name` path not to exist. I want to create name sub directory only if it doesn't exist. `os.MkdirAll` wont return an error if dir path exists which defeats the purpose.

vdemeester replied:
@shashwathi ah good point :sweat_smile:

fejta replied:
What if I want to check out the `k8s.io/test-infra` source to `/workspace/go/src/k8s.io/test-infra` rather than just `/workspace/test-infra`?

This is often an important use case for go programs.

pivotal-nader-ziada commented on pkg/builder/cluster/convert/convert.go:
```diff
@@ -132,21 +140,32 @@ func containerToGit(git corev1.Container) (*v1alpha1.SourceSpec, error) {
+	name := ""
+	if len(git.Args) == 6 {
+		name = git.Args[5]
+	}
```
can you add a comment to what this Args[5] is?

shashwathi replied:
There is a lot of to and from conversion confusion in this code. I will add more comments to explain what is happening.

shashwathi commented on pkg/apis/build/v1alpha1/build_types.go:
```diff
@@ -60,6 +60,9 @@ type BuildSpec struct {
 	Source *SourceSpec `json:"source,omitempty"`
 
+	// Sources specifies the inputs to the build.
+	Sources []*SourceSpec `json:"sources,omitempty"`
```
Looks like most of the types did not have `optional` tag. I will update them

shashwathi commented:
@pivotal-nader-ziada : I have addressed your comments. Let me know if comments in code cleared up some confusion but if not I will take a stab at adding more comments

fejta commented on pkg/builder/cluster/convert/convert.go:
```diff
@@ -105,48 +105,76 @@ var (
+		Image:        *gitImage,
+		Args:         args,
 		VolumeMounts: implicitVolumeMounts,
 		WorkingDir:   workspaceDir,
```
If every source container uses the same working dir will multiple sources clobber each other?

fejta replied:
Nevermind! the answer here is clearly no :)

fejta commented on pkg/apis/build/v1alpha1/build_types.go:
```diff
@@ -137,20 +149,28 @@ type ArgumentSpec struct {
 	SubPath string `json:"subPath,omitempty"`
+
+	// Name is the path under which source will be mounted.
+	// +optional
```
It looks like this is only optional in `source`, not `sources`. We should document this.

fejta commented on pkg/builder/cluster/convert/convert.go:
```diff
@@ -464,25 +527,29 @@ func ToCRD(pod *corev1.Pod) (*v1alpha1.Build, error) {
+	var scm []*v1alpha1.SourceSpec
+	for _, step := range steps {
+		conv := stepNameConversionToSource(step.Name)
+		if conv != nil {
```
conv is not used outside the scope of this if statement, please combine this line with the prev one `if conv := sNCTS(); conv != nil {`

fejta commented on pkg/apis/build/v1alpha1/build_types.go:
```diff
@@ -137,20 +149,28 @@ type ArgumentSpec struct {
+	// +optional
 	SubPath string `json:"subPath,omitempty"`
+
+	// Name is the path under which source will be mounted.
```
So this means that if I specify `source: {name: foo, ...}` then my steps can expect there to be a foo directory in `$PWD` aka `/workspace`?

nit: consider something like `// Name is the subpath in workspace where the source will be copied`. It doesn't seem like "mount" is the correct word here as all sources are copied into the workspace volumeMount, rather than a unique volumeMount per source.

fejta commented on cmd/git-init/main.go:
```diff
@@ -66,10 +68,30 @@ func main() {
+
+	if *name != "" {
+		// create dir name
+		path := filepath.Join(dir, *name)
```
This doesn't seem to be doing anything... How is `os.Stat(path)` different from `os.Stat(*name)`?

fejta commented on pkg/apis/build/v1alpha1/build_validation.go:
```diff
@@ -92,3 +96,38 @@ func (bt *BuildSpec) validateTimeout() *apis.FieldError {
+			if _, ok := names[source.Name]; ok {
+				return apis.ErrMultipleOneOf("b.spec.sources.names")
+			}
+			// multiple sources cannot have subpath defined
```
Nice!

fejta commented on pkg/builder/cluster/convert/convert.go:
```diff
@@ -105,48 +105,76 @@ var (
+	// source name is empty then use `build-step-gcs-source` name
+	containerName := initContainerPrefix + gcsSource
+
+	if name != "" {
```
Also add the `--name` flag?

shashwathi replied:
README does not document `dest_dir`. I looked into codebase to get that flag. I will update it here and e2e test. Thanks fir catching that.

fejta commented on pkg/apis/build/v1alpha1/build_types.go:
```diff
@@ -137,20 +149,28 @@ type ArgumentSpec struct {
+
+	// Name is the path under which source will be mounted.
+	// +optional
+	Name string `json:"name,omitempty"`
```
Since this gets appended into the container name, we should note the restrictions on allowed characters:
* Must be a basename (no /)
* Must be a [valid DNS name](https://stackoverflow.com/questions/27791913/what-characters-are-allowed-in-kubernetes-port-and-container-names) (only alphanumeric characters, no _ or -)

Also would "Path", "TargetSubPath" etc be a better name here?

shashwathi replied:
- If Path or TagetPath is used then it should allow `/` character.  
- I thought `Name` would be a good option because its unique identity of source.

I do like the TargetPath feature but I think it is separate feature and I am not sure about repurposing `Name` to be targetPath. I understand the use case and I think it is completely valid

fejta replied:
Once/if we have TargetPath what is the purpose of Name?

I think we should consider replacing Name with TargetPath as this seems like an important feature because it is an essential feature for go projects. And go is an important language for knative and kubernetes.

Including `TargetPath` support seems more important than the `WorkDir` (cloning to /workspace) conversation below.

The workaround here would be to add a step which moves each source from /workspace/foo to /workspace/go/src/whatever.domain/path/to/package, which may be complicated to extract from "foo" (especially since foo can't contain most obvious separate chars)

shashwathi replied:
You have valid points. 
- I am misusing `Name` as TargetPath in this PR.  I will rename the variable to `TargetPath`. Question: Is  there value in adding validation to check if multiple sources don't have same `TargetPath` or is that too much validation in build spec ?

- I think there is still value in having `Name` field but the purpose is purely for logging, generating init container names. Init containers with meaningful names made my life easy when debugging. Make the field optional for user so less things to declare I guess. WDYT?

fejta commented on test/sources/test3.yaml:
```diff
@@ -0,0 +1,36 @@
+  - name: "gcs-source"
+    gcs:
+      type: Archive
+      location: gs://build-crd-tests/rules_docker-master.zip
```
Do these tests run on presubmit?

shashwathi replied:
Yes. presubmit runs both e2e go tests and yaml tests

fejta commented on cmd/git-init/main.go:
```diff
@@ -66,10 +68,30 @@ func main() {
+		logger.Errorf("Failed to get current dir", err)
+	}
+
+	if *name != "" {
```
This all seems largely equivalent to:

```golang
if len(*name) > 0 {
  run(logger, "git", "init", *name)
  os.Chdir(*name)
} else {
  run(logger, "git", "init")
}
```

shashwathi replied:
yes agreed 100%. I like this better and its so simple :D

fejta commented on pkg/apis/build/v1alpha1/build_validation.go:
```diff
@@ -92,3 +96,38 @@ func (bt *BuildSpec) validateTimeout() *apis.FieldError {
+	for _, source := range bs.Sources {
+		// check all source have names
+		if source != nil {
+			if source.Name == "" {
```
Does it make sense to allow checking out one source (probably the first) to /workspace root (not sure the answer)?

shashwathi replied:
I did not consider this use case. I guess you want to use `sources` but want the same behavior as `source` where it is present in `/workspace` ? (Just to make sure I understood your question correctly)

Great question. I do not have strong opinion about continuing to use `source` or support this feature in `sources`. Do you have an opinion about this

fejta replied:
I don't have an opinion about this, but thought I'd mention it because builds seem to generally assume that PWD has the primary source in it. So I suspect it will be fairly common to have your primary source (aka the repo with the PR that needs to be tested) plus any additional repos the test needs to include. 

If I start out with a single repo and then need to add a second repo later, currently I'll need to update all the steps to set `WorkDir: /workspace/first-repo-name`.

The way our cloning container handles this is by special casing the "primary" ref and defaulting the working directory to this /workspace/go/src/org/repo subpath rather than /workspace 
 (see https://github.com/kubernetes/test-infra/pull/9907)

So it seems like it would be nice if we either allow checking out a repo to /workspace or else specifying which source I want to set my workdir to.

shashwathi replied:
If we use `TargetPath` field and if that field is not set then source will be present in `/workspace`. 
One problem I see with this approach is build with 2 sources and no targetPaths will dump contents in `/workspace` and if there is any clash of file names, user will not realize this I suppose.

shashwathi replied:
Nevermind. This will be be an error

shashwathi commented:
/hold

shashwathi commented on pkg/builder/cluster/convert/convert.go:
```diff
@@ -257,31 +307,41 @@ func FromCRD(build *v1alpha1.Build, kubeclient kubernetes.Interface) (*corev1.Po
 		}
-
-		workspaceSubPath = build.Spec.Source.SubPath
+		workspaceSubPath = source.SubPath
```
I have added comment on line 346 so I am marking this as resolved

shashwathi commented on cmd/git-init/main.go:
```diff
@@ -66,10 +68,26 @@ func main() {
 	}
 
-	run(logger, "git", "init")
+	dir, err := os.Getwd()
```
I think so. You can re-use git-init in build steps and provide different workspaceDir. I am also not fan of making assumptions about directory structure. Its better to be explicit in the code.

shashwathi commented:
>  If it's empty what will this be? A hash? Index?

Index. Might be easier to associate when debugging.

> Also should TargetPath /foo be allowed or should this have to be a path inside /workspace (so /workspace/foo)?

`TargetPath` will be appended to `/workspace`. 

> Validating that TargetPath is unique sounds reasonable to me (not bothering yet also seems reasonable). How about source A's TargetPath being /foo/a and source B's being /foo/a/b? Do you want to allow this or require sources to not go inside another source?

- Only allow unique TargetPath 
- Not allow nested TargetPath

Goal at end of fetching sources should be deterministic directory structure. Steps will allow to mess around with this structure anyway after.

WDYT? @fejta

shashwathi commented:
/test pull-knative-build-integration-tests

fejta commented:
All SGTM

shashwathi commented:
/hold cancel

shashwathi commented:
@fejta : I have updated the PR. Please review it and let me know if you have any feedback

ImJasonH commented on pkg/apis/build/v1alpha1/build_types.go:
```diff
@@ -113,14 +122,17 @@ type TemplateInstantiationSpec struct {
 
 	// Arguments, if specified, lists values that should be applied to the
 	// parameters specified by the template.
+	// +optional
```
Can we make adding `+optional` a separate later PR?

Does anything enforce that non-`+optional` fields are specified? Or is it just for documentation purposes?

shashwathi replied:
I am okay with separate PR for `+optional`. I will add `optional` tags for fields introduced in this PR only.

> Does anything enforce that non-+optional fields are specified? 

k8s api should be enforcing them in future. 

Ref: https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#optional-vs-required

shashwathi replied:
https://github.com/knative/build/pull/453

ImJasonH commented on pkg/apis/build/v1alpha1/build_validation.go:
```diff
@@ -92,3 +97,57 @@ func (bt *BuildSpec) validateTimeout() *apis.FieldError {
+func (bs BuildSpec) validateSources() *apis.FieldError {
+	var subPathExists bool
+	names := map[string]string{}
+	var sourceExists bool
```
I don't think we gain much from making this a variable.

Line 112 could just be:

```
if len(bs.Sources) > 0 && b.Source != nil {
 // boom
}
```

ImJasonH commented on pkg/apis/build/v1alpha1/build_validation.go:
```diff
@@ -18,6 +18,7 @@ package v1alpha1
 
```
Should we have a mutating webhook modify incoming builds that specify `Source` to become a build that specifies a single `Sources`? From then on we can assume/require builds don't have `Source` specified. Then in the future we can remove the ability to specify `Source` at all.

The downside of relying on a webhook is that some providers don't support them, so we'd have to duplicate the check in the controller. :(

shashwathi replied:
> Should we have a mutating webhook modify incoming builds that specify Source to become a build that specifies a single Sources?

Sounds like a great idea. 

>The downside of relying on a webhook is that some providers don't support them, so we'd have to duplicate the check in the controller. :(

How about continuing support for couple of minor versions before dropping support for `Source` in builds altogether? 

Do we need to address this in this PR?

ImJasonH commented on pkg/builder/cluster/convert/testdata/gcs-archive-sources.yaml:
```diff
@@ -0,0 +1,27 @@
+  targetPath: "workspace/foo"
+steps:
+- name: helloworld
+  image: gcr.io/google-containers/busybox
```
If this doesn't actually have to be a real image, since it's not actually going to run anything, this could just be:

```
steps:
- image: hello
```

shashwathi replied:
Sure, I will update this file. I think there are examples in testdata that declare `gcr.io/google-containers/busybox` as image. 
ref: https://github.com/knative/build/blob/master/pkg/builder/cluster/convert/testdata/gcs-archive.yaml#L20

ImJasonH commented on test/e2e/simple_test.go:
```diff
@@ -475,3 +475,136 @@ func TestPersistentVolumeClaim(t *testing.T) {
+
+	// Emit a metric for null-build latency (i.e., time to schedule and execute
+	// and finish watching a build).
+	_, span := trace.StartSpan(context.Background(), "NullBuildLatency")
```
Remove this, since we should only report this metric for the simple successful build test.

ImJasonH commented on test/e2e/simple_test.go:
```diff
@@ -475,3 +475,136 @@ func TestPersistentVolumeClaim(t *testing.T) {
+
+	// Emit a metric for null-build latency (i.e., time to schedule and execute
+	// and finish watching a build).
+	_, span := trace.StartSpan(context.Background(), "NullBuildLatency")
```
ditto

shashwathi commented:
/test pull-knative-build-integration-tests

shashwathi commented:
/test pull-knative-build-integration-tests

fejta commented on pkg/apis/build/v1alpha1/build_validation.go:
```diff
@@ -92,3 +97,53 @@ func (bt *BuildSpec) validateTimeout() *apis.FieldError {
+	if _, ok := pathMap[path]; ok {
+		return err
+	}
+	for key := range pathMap {
```
Seems like we can put a `if path == "" { return nil }` before this? 

What behavior do you want when TargetPath == ""?

fejta commented on pkg/builder/cluster/convert/convert.go:
```diff
@@ -105,87 +106,166 @@ var (
+	args := []string{"-url", git.Url,
+		"-revision", git.Revision,
+	}
+	containerName := initContainerPrefix + gitSource
```
nit: may want to move this after finishing with the args now

shashwathi commented on pkg/apis/build/v1alpha1/build_validation.go:
```diff
@@ -92,3 +97,53 @@ func (bt *BuildSpec) validateTimeout() *apis.FieldError {
+		return err
+	}
+	for key := range pathMap {
+		if path != "" && strings.HasPrefix(key, path) {
```
Good catch. thanks 👍

shashwathi commented:
> aka does this happen in pull-knative-build-integration-tests? or which one?

Yes `pull-knative-build-integration-tests`

shashwathi commented on pkg/apis/build/v1alpha1/build_validation.go:
```diff
@@ -92,3 +97,53 @@ func (bt *BuildSpec) validateTimeout() *apis.FieldError {
+}
+
+// validatePath checks that source path is unique
+// or subpath is unique
```
I agree `subPath` is confusing in this context. I will update the comment

fejta commented on cmd/git-init/main.go:
```diff
@@ -66,10 +68,26 @@ func main() {
+	}
+
+	if *path != "" {
+		runOrFail(logger, "git", "init", *path)
```
Nice looks like `git init foo/bar/whatever` works!

fejta commented on pkg/apis/build/v1alpha1/build_validation.go:
```diff
@@ -92,3 +96,43 @@ func (bt *BuildSpec) validateTimeout() *apis.FieldError {
+
+		if source.TargetPath == "" {
+			if emptyTargetPath {
+				return apis.ErrMultipleOneOf("b.spec.sources.targetPath")
```
can't we provide a more helpful error message here? Seems similar to subpath

shashwathi replied:
Invalid value makes more sense here. I will update

shashwathi commented:
@fejta @ImJasonH : I have updated the logic for targetPath validation since the previous one was fragile. Please review the PR and leave comments.

fejta commented on test/e2e/invalid_test.go:
```diff
@@ -78,6 +78,72 @@ func TestInvalidBuild(t *testing.T) {
+	}, {
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: buildTestNamespace,
+			Name:      "source-with-duplicate-subpath",
```
I don't understand the problem with this testcase.

I would expect specifying two sources, one going to `paht/foo` and the other `path/foo` should work since `paht` != `path`

fejta commented on test/e2e/invalid_test.go:
```diff
@@ -78,6 +78,72 @@ func TestInvalidBuild(t *testing.T) {
+		},
+		Spec: v1alpha1.BuildSpec{
+			Sources: []v1alpha1.SourceSpec{{
+				TargetPath: "paht/foo",
```
Is this a typo for path? If not can you choose a name that's more distinct from path?

shashwathi replied:
I will fix that.

fejta commented on pkg/builder/cluster/convert/convert.go:
```diff
@@ -105,87 +106,169 @@ var (
 }
 
-func customToContainer(source *corev1.Container) (*corev1.Container, error) {
+func customToContainer(source *corev1.Container, name string) (*corev1.Container, error) {
```
How are custom containers supposed to consume the value of `targetPath`? 

It seems like we should either inject the value somehow (into an env?) or else reject targetPath != ""

shashwathi replied:
Thats a valid point. Currently targetPath is not used in custom sources so we should raise and document it in targetPath.

shashwathi replied:
Fixed it and added test case for this.

fejta commented on pkg/apis/build/v1alpha1/build_validation_test.go:
```diff
@@ -38,6 +38,369 @@ func TestValidateBuild(t *testing.T) {
+			Spec: BuildSpec{
+				Sources: []SourceSpec{{
+					Name:       "gitpathab",
+					TargetPath: "/a/b/d",
```
Should this error? What's the expected result of this?

shashwathi replied:
I moved all path related test cases into target_path_test.go. Let me know if I covered this case

fejta commented on pkg/apis/build/v1alpha1/build_validation_test.go:
```diff
@@ -38,6 +38,369 @@ func TestValidateBuild(t *testing.T) {
+			},
+		},
+	}, {
+		desc: "sources with duplicate targetPath",
```
I'd really like to see us validate the core use case of checking out two github repos to their correct gopath location, aka something like:

```golang
build: &Build{
	Spec: BuildSpec{
		Sources: []SourceSpec{{
			TargetPath: "/go/src/github.com/knative/build",
			Git: &GitSourceSpec{
				Url:      "https://github.com/knative/build",
				Revision: "revision",
			},
		}, {
			TargetPath: "/go/src/github.com/knative/serving",
			Git: &GitSourceSpec{
				Url:      "https://github.com/knative/serving",
				Revision: "revision",
			},
		}},
		Steps: []corev1.Container{{
			Name:  "foo",
			Image: "gcr.io/foo-bar/baz:latest",
		}},
	},
},
reason: "sources with prefixed targetPath",
```

It isn't clear to me that this scenario will work since they both duplicate the /go/src/github.com/knative prefix

shashwathi replied:
That should work as the leaf dir is different.

fejta commented on pkg/apis/build/v1alpha1/target_path_validation.go:
```diff
@@ -0,0 +1,61 @@
+// insert function breaks the path and inserts every element into nodeMap
+// insert function throws error if the path is already traversed by any
+// previous targetPaths
+func insertNode(path string, nodeMap map[string]*node) *apis.FieldError {
```
Why does this one function need its own file?

Can we get a TestInsertNode unit test in `target_path_validation_test.go`?

I'm not 100% clear this works correctly (seems like it will only allow a single source to have a github.com prefix? Which is definitely not what we want. We just want to make sure that A is not a subset of B and B is not a subset of A, but it is perfectly fine to have A and B both be subsets of C so long as C is not a source, aka

OK:
A: github.com/foo/bar
B: github.com/foo/frobber

Not ok:
A: github.com/foo
B: github.com/foo/bar

Not ok:
A: github.com/foo/bar
B: github.com/foo

I think this method is trying to raise an error if any two paths start with github.com/foo, right?

shashwathi replied:
True. I missed some use cases. Let me fix it and ping you on this thread

shashwathi replied:
@fejta : I have updated the PR to include the above test cases.

shashwathi commented:
/hold

shashwathi commented:
/hold cancel

fejta commented on pkg/apis/build/v1alpha1/target_path_validation_test.go:
```diff
@@ -0,0 +1,151 @@
+		},
+		wantErr: apis.ErrMultipleOneOf("b.spec.sources.targetPath"),
+	}, {
+		desc: "paths with same length and different leaf node",
```
Yay!

shashwathi commented on test/sources/test3.yaml:
```diff
@@ -0,0 +1,37 @@
+# limitations under the License.
+apiVersion: build.knative.dev/v1alpha1
+kind: Build
+metadata:
```
I will submit a PR with additional yaml tests for this.
